wallet/src/repository/wallet/mod.rs

```diff
@@ -514,123 +520,172 @@ where
-                input_amount = input_amount
-                    .checked_add(amount)
-                    .ok_or_else(|| Error::TransactionBalanceOverflow)?;
-            }
-        }
-
-        let mut output_amount: u64 = 0;
-        for (index, output) in txn.outputs.iter().enumerate() {
-            if let Some(model::Path { .. }) = self.db.get_opt(&keys::pkh(&output.pkh))? {
-                let address = output.pkh.bech32(if self.params.testnet {
-                    Environment::Testnet1
-                } else {
-                    Environment::Mainnet
-                });
-                let key_balance = model::KeyBalance {
-                    pkh: output.pkh,
-                    amount: output.value,
-                };
-
-                let out_ptr = model::OutPtr {
-                    txn_hash: Vec::from(txn_hash),
-                    output_index: index as u32,
-                };
-
-                match db_utxo_set.insert(out_ptr.clone(), key_balance.clone()) {
-                    Some(_) => {
-                        log::warn!(
-                            "Found an already existing transaction to our address {}! Output pointer: {:?}",
-                            address,
-                            out_ptr
-                        );
-                    }
-                    _ => {
-                        log::info!(
-                            "Found transaction to our address {}! Amount: +{} nanowits",
-                            address,
-                            output.value
-                        );
-                    }
-                }
-                state.utxo_set.insert(out_ptr, key_balance);
-
-                output_amount = output_amount
-                    .checked_add(output.value)
-                    .ok_or_else(|| Error::TransactionBalanceOverflow)?;
-            }
-        }
-
-        let txn_balance = (output_amount as i128)
-            .checked_sub(input_amount as i128)
-            .ok_or_else(|| Error::TransactionBalanceUnderflow)?;
+        // For each type of transaction, get mutations and transaction balance
+        let transaction_mutation = match txn {
+            types::Transaction::DataRequest(txn) => self._check_txn_balance(
+                &state.utxo_set,
+                txn.hash().as_ref(),
+                &txn.body.inputs,
+                &txn.body.outputs,
+            )?,
+            types::Transaction::ValueTransfer(txn) => self._check_txn_balance(
+                &state.utxo_set,
+                txn.hash().as_ref(),
+                &txn.body.inputs,
+                &txn.body.outputs,
+            )?,
+            _ => return Err(Error::UnsupportedTransactionType(format!("{:?}", txn))),
+        };
 
-        let txn_balance =
-            i64::try_from(txn_balance).map_err(|_| Error::TransactionBalanceOverflow)?;
+        // If balance changes, then update memory state and DB
+        if transaction_mutation.txn_balance != 0 {
```

Comment:
Is this correct? What if the transaction has zero balance but spends some UTXOs? For example sending a transaction to ourselves.

Reply:
Good catch!